Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -1372,7 +1381,17 @@ def visit_WithStatNode(self, node):
                     name, value = directive
                     if name in ('nogil', 'gil'):
                         # special case: in pure mode, "with nogil" spells "with cython.nogil"
-                        node = Nodes.GILStatNode(node.pos, state = name, body = node.body)
+                        condition = None
+                        if isinstance(node.manager, ExprNodes.SimpleCallNode) and len(node.manager.args) > 0:
+                            if len(node.manager.args) == 1:
+                                condition = node.manager.args[0]
+                            else:
+                                self.context.nonfatal_error(
+                                    PostParseError(node.pos, "Compiler directive %s accepts one positional argument." % name))
+                        elif isinstance(node.manager, ExprNodes.GeneralCallNode):
+                            self.context.nonfatal_error(
+                                PostParseError(node.pos, "Compiler directive %s accepts one positional argument." % name))
+                        node = Nodes.GILStatNode(node.pos, state = name, body = node.body, condition=condition)
```

Comment:
Let's clean up the code while we're at it.
```suggestion
                        node = Nodes.GILStatNode(node.pos, state=name, body=node.body, condition=condition)
```